Mathlib/CategoryTheory/Preadditive/HomOrthogonal.lean

```diff
@@ -0,0 +1,238 @@
+    ((⨁ fun a => s (f a)) ⟶ ⨁ fun b => s (g b)) ≃+
+      ∀ i : ι, Matrix (g ⁻¹' {i}) (f ⁻¹' {i}) (End (s i)) :=
+  { o.matrixDecomposition with
+    map_add' := fun w z => by
+      ext
+      dsimp [biproduct.components]
+      simp }
+#align category_theory.hom_orthogonal.matrix_decomposition_add_equiv CategoryTheory.HomOrthogonal.matrixDecompositionAddEquiv
+
+@[simp]
+theorem matrixDecomposition_id (o : HomOrthogonal s) {α : Type} [Fintype α] {f : α → ι} (i : ι) :
+    o.matrixDecomposition (𝟙 (⨁ fun a => s (f a))) i = 1 := by
+  ext (⟨b, ⟨⟩⟩⟨a, j_property⟩)
+  simp only [Set.mem_preimage, Set.mem_singleton_iff] at j_property
+  simp only [Category.comp_id, Category.id_comp, Category.assoc, End.one_def, eqToHom_refl,
+    Matrix.one_apply, HomOrthogonal.matrixDecomposition_apply, biproduct.components]
+  split_ifs with h
+  · cases h
+    simp
+  · simp at h
+    -- porting note: used to be `convert comp_zero`, but that does not work anymore
+    have : biproduct.ι (fun a ↦ s (f a)) a ≫ biproduct.π (fun b ↦ s (f b)) b = 0 := by
+      simpa using biproduct.ι_π_ne _ (Ne.symm h)
+    rw [this, comp_zero]
+
+#align category_theory.hom_orthogonal.matrix_decomposition_id CategoryTheory.HomOrthogonal.matrixDecomposition_id
+
+theorem matrixDecomposition_comp (o : HomOrthogonal s) {α β γ : Type} [Fintype α] [Fintype β]
+    [Fintype γ] {f : α → ι} {g : β → ι} {h : γ → ι} (z : (⨁ fun a => s (f a)) ⟶ ⨁ fun b => s (g b))
+    (w : (⨁ fun b => s (g b)) ⟶ ⨁ fun c => s (h c)) (i : ι) :
+    o.matrixDecomposition (z ≫ w) i = o.matrixDecomposition w i ⬝ o.matrixDecomposition z i := by
+  ext (⟨c, ⟨⟩⟩⟨a, j_property⟩)
+  simp only [Set.mem_preimage, Set.mem_singleton_iff] at j_property
+  simp only [Matrix.mul_apply, Limits.biproduct.components,
+    HomOrthogonal.matrixDecomposition_apply, Category.comp_id, Category.id_comp, Category.assoc,
+    End.mul_def, eqToHom_refl, eqToHom_trans_assoc, Finset.sum_congr]
+  conv_lhs => rw [← Category.id_comp w, ← biproduct.total]
+  simp only [Preadditive.sum_comp, Preadditive.comp_sum]
+  apply Finset.sum_congr_set
+  · intros
+    simp
+  · intro b nm
+    simp only [Set.mem_preimage, Set.mem_singleton_iff] at nm
+    simp only [Category.assoc]
+    -- porting note: this used to be 4 times `convert comp_zero`
+    have : biproduct.ι (fun a ↦ s (f a)) a ≫
+      z ≫ biproduct.π (fun b ↦ s (g b)) b ≫ biproduct.ι (fun b ↦ s (g b)) b ≫ w ≫
+      biproduct.π (fun b ↦ s (h b)) c  = 0 := by
+      have : z ≫ biproduct.π (fun b ↦ s (g b)) b ≫ biproduct.ι (fun b ↦ s (g b)) b ≫ w ≫
+          biproduct.π (fun b ↦ s (h b)) c = 0 := by
+        have : biproduct.π (fun b ↦ s (g b)) b ≫ biproduct.ι (fun b ↦ s (g b)) b ≫ w ≫
+            biproduct.π (fun b ↦ s (h b)) c = 0 := by
+          have : biproduct.ι (fun b ↦ s (g b)) b ≫ w ≫ biproduct.π (fun b ↦ s (h b)) c = 0 := by
+            apply o.eq_zero nm
+          rw [this, comp_zero]
+        rw [this, comp_zero]
+      rw [this, comp_zero]
+    rw [this, comp_zero]
```

Comment:
needs some golfing

Reply:
I tried `refine' comp_zero`, even providing some of the arguments, but no luck. Also `convert` does not seem up for the task anymore, even when I try to get more explicit with `convert @comp_zero ...`

Reply:
perhaps it is fine to leave it as is with the fix, accepting that `convert` is now a little weaker...